src/hotspot/share/opto/escape.cpp

```diff
@@ -401,6 +416,239 @@ bool ConnectionGraph::compute_escape() {
   return has_non_escaping_obj;
 }
 
+// Check if it's profitable to reduce the Phi passed as parameter.  Returns true
+// if at least one scalar replaceable allocation participates in the merge and
+// no input to the Phi is nullable.
+bool ConnectionGraph::can_reduce_this_phi_inputs(PhiNode* phi) const {
+  // Check if there is an scalar replaceable allocate in the Phi
+  bool found_sr_allocate = false;
+
+  for (uint i = 1; i < phi->req(); i++) {
+    // Right now we can't restore a "null" pointer during deoptimization
+    if (_igvn->type(phi->in(i))->maybe_null()) {
+      NOT_PRODUCT(if (TraceReduceAllocationMerges) tty->print_cr("Can NOT reduce Phi %d. Input %d is nullable.", phi->_idx, i);)
+      return false;
+    }
+
+    // We are looking for at least one SR object in the merge
+    JavaObjectNode* ptn = unique_java_object(phi->in(i));
+    if (ptn != NULL && ptn->scalar_replaceable()) {
+      assert(ptn->ideal_node() != NULL && ptn->ideal_node()->is_Allocate(), "sanity");
+      Node* alloc = ptn->ideal_node()->as_Allocate();
+
+      if (PhaseMacroExpand::can_eliminate_allocation(_igvn, alloc->as_Allocate(), NULL, true)) {
```

Comment:
Do we need to call as_Allocate() again, since it was already called on line 437?

Reply:
I can reuse this. Thanks.